Add ms:format-date and ms:format-time test cases

IDisposable opened the pull request:
Test cases for regression bug #93189 (difference between 4.52 and 6.0)

Currently set to _PASS_ with the new (broken) values emitted by .Net 6.0+ 

When #97402 is landed, we need to update the expected values per [this comment](https://github.com/dotnet/runtime/pull/97402#issuecomment-2749607737)

IDisposable committed:
Add ms:format-date and ms:format-time test cases

Test cases for regression bug #93189 (difference between 4.52 and 6.0)

NOTE: The baseline\bug9389.xml reflects the _current_ (broken) behavior so these tests will currently be PASSING. When the bug is resolved, this test data needs to be updated to reflect what _actually_ should happen.

Fix problem with writing out the diff.xml file when test fails
You have to open the file in Create mode with Write access.

Turns out that the separator between the minutes and the AM/PM is not a space, it's actually a n U+202F NARROW NO-BREAK SPACE so update the expected value.

Add code to normalize spaces in the XmlDiff. Since it's hard to see different kinds of spaces, add the option to treat them all as normal spaces.

IDisposable commented:
When https://github.com/dotnet/runtime/pull/97402 is landed, we need to update the expected values in
_D:\Dev\dotnet\runtime\src\libraries\System.Private.Xml\tests\Xslt\TestFiles\TestData\XsltApiV2\baseline\bug93189.xml_ should change:

- out/dates/test3 to be `2009-06-15` 
- out/times/test3 to be `13:45`

IDisposable commented:
Updated the XmlDiff classes to handle space-normalization (e.g. various things like non-breaking space, zero-width space, etc.) to normal space to ease test case baseline authoring. This is due to the fact that on the test environment, the time format is AM/PM, and turns out that the separator between the minutes and the AM/PM is not a space, it's actually a n U+202F NARROW NO-BREAK SPACE so the emitted actual wouldn't match the expected unless you knew to paste that exact character in.

IDisposable commented:
@danmoseley I added a test for PR #97402 [per your request](https://github.com/dotnet/runtime/pull/97402#issuecomment-1906893875)

IDisposable commented:
Tagging folks that might want to review this since the auto-tagger didn't.

XML folks @jeffhandley @StephenMolloy @HongGit 

PR #97402 reviewers @tarekgh @krwq

tarekgh commented on src/libraries/Common/tests/System/Xml/XmlDiff/XmlDiff.cs:
```diff
@@ -437,6 +438,20 @@ private DiffType CompareNodes(XmlDiffNode sourceNode, XmlDiffNode targetNode)
+                return value;
+
+            if (NormalizeSpaces)
+                value = Regex.Replace(value, "[\u00A0\u180E\u2000-\u200B\u202F\u205F\u3000\uFEFF]", " ");
```
>"[\u00A0\u180E\u2000-\u200B\u202F\u205F\u3000\uFEFF]" [](http://example.com/codeflow?start=45&length=53)

From where did you get this list? 

`\u180E` is a format character and not really a space. `\uFEFF` similar. And missing `\u1680`.

[https://unicodeplus.com/category/Zs](https://unicodeplus.com/category/Zs)

IDisposable replied:
The list I used is [Unicode spaces](https://jkorpela.fi/chars/spaces.html) pointed at by this [answer](https://stackoverflow.com/a/51399843/2076) and it is based on the [Unicode category Zs](https://www.compart.com/en/unicode/category/Zs).

It adds [\FEFF is zero-width no-break space](https://www.compart.com/en/unicode/U+FEFF) and seems to be missed by a lot of folks.

It adds [\u180E Mongolian vowel separator](https://www.compart.com/en/unicode/U+180E) because it is treated as a space, happy to remove it.

I removed [\u1680-Ogham space mark](https://www.compart.com/en/unicode/U+1680) because its glyph is usually rendered as a dash,  could easily put it back.

Happy to adjust this pattern as desired.

IDisposable replied:
It's not entirely about XML's view of spaces, this whole change is motivated because in the test-environment's locale the format for times is `h:mm AM` except the space _isn't actually a space_ but is really the [\u202F Narrow No-Break Space](https://unicodeplus.com/U+202F) which broke my naive test. When the need for a normalization to spaces (to defend against locale-driven possibilities) I figured I should cast a slightly wider net.
 
There's absolutely no objection in my mind to just using `\s` in the pattern since that only result in ignoring a few more 
"blanks" and this is nothing but test code anyway. Point me and I'll shoot as ordered :)

tarekgh replied:
In the date/time parsing code we just call [`char.IsWhiteSpace`](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeParse.cs,5486). Either you continue using regex as you suggested or just manually create a helper method iterate over the string checking using `char.IsWhiteSpace`. What ever you feel easier to you, do it :-)

Thanks for your help fixing the tests!

tarekgh replied:
one note, if you use the regex, most likely you need `\p{Zs}` instead of `\s`. I don't recall if `\s` cover all different whitespaces or not.

tarekgh commented on src/libraries/Common/tests/System/Xml/XmlDiff/XmlDiffDocument.cs:
```diff
@@ -1550,15 +1560,19 @@ public class XmlDiffCharacterData : XmlDiffNode
+                value = value.Replace("\n", "").Replace("\r", "");
+
+            if (NormalizeSpaces)
+                value = Regex.Replace(value, "[\u00A0\u180E\u2000-\u200B\u202F\u205F\u3000\uFEFF]", " ");
```
> value = Regex.Replace(value, "[\u00A0\u180E\u2000-\u200B\u202F\u205F\u3000\uFEFF]", " "); [](http://example.com/codeflow?start=15&length=90)

would it be better to move this to a helper method instead of duplicating it?

IDisposable replied:
Certainly, or a Regex member, or at least a constant :) Good call.

What's the preference for this sort of thing in test classes?

tarekgh commented on src/libraries/Common/tests/System/Xml/XmlDiff/XmlDiffDocument.cs:
```diff
@@ -1550,15 +1560,19 @@ public class XmlDiffCharacterData : XmlDiffNode
-                this._value = this._value.Replace("\n", "");
-                this._value = this._value.Replace("\r", "");
-            }
+                value = value.Replace("\n", "").Replace("\r", "");
```
>"\n" [](http://example.com/codeflow?start=38&length=4)

use character instead of string here? `'\n'` and `'\r'`

IDisposable replied:
We're* completely stripping the CR and LF out entirely, can't do a `String.Replace(char, char)` here can we? What would the replacement character be?

(*this is preexisting code)

tarekgh replied:
never mind, somehow, I saw the empty string as space.

tarekgh commented:
> // The .NET Foundation licenses this file to you under the MIT license.

not seeing any change in this file 

---
Refers to: src/libraries/System.Private.Xml/tests/Xslt/XslTransformApi/CXslTransform.cs:2 in 5dd8d51. [](commit_id = 5dd8d51c5e4725c21856f98fd03a226bc055ba0f, deletion_comment = False)

tarekgh commented on src/libraries/Common/tests/System/Xml/XmlDiff/XmlDiffDocument.cs:
```diff
@@ -1550,15 +1560,19 @@ public class XmlDiffCharacterData : XmlDiffNode
         private string _value;
         private XmlDiffNodeType _nodetype;
-        public XmlDiffCharacterData(string value, XmlDiffNodeType nt, bool NormalizeNewline)
+        public XmlDiffCharacterData(string value, XmlDiffNodeType nt, bool NormalizeNewline, bool NormalizeSpaces, bool IgnoreWhitespace)
```
> NormalizeSpaces, bool IgnoreWhitespace [](http://example.com/codeflow?start=97&length=39)

it is weird we are using parameters started with upper case letter!

IDisposable replied:
Changed, was just following the pattern.

tarekgh reviewed (approved):
Left minor comments. LGTM, otherwise. 

@krwq could you have a quick look too?

IDisposable commented:
> > // The .NET Foundation licenses this file to you under the MIT license.
> 
> not seeing any change in this file
> 
> Refers to: src/libraries/System.Private.Xml/tests/Xslt/XslTransformApi/CXslTransform.cs:2 in [5dd8d51]

There was a trailing space that VSCode stripped while I churning on this (ended up not making any other changes to this files). Would you like me to take it out of the commits?

https://github.com/dotnet/runtime/pull/113507/files#diff-4a94ae4a2848c89aa86609a761b9384b64e292a1f48c3218d0a6ec384feae87cL904

IDisposable committed:
PR feedback

Moved the space-replacements pattern to XmlDiffOption.
Lowercased the constructor arguments for XmlDiffCharacterData.
Make XmlDiffProcessingInstruction pass options as named parameters.

tarekgh commented:
> There was a trailing space that VSCode stripped while I churning on this (ended up not making any other changes to this files). Would you like me to take it out of the commits?

That is fine to keep it. I was only wondering as was not seeing real changes.

IDisposable commented:
> > There was a trailing space that VSCode stripped while I churning on this (ended up not making any other changes to this files). Would you like me to take it out of the commits?
> 
> That is fine to keep it. I was only wondering as was not seeing real changes.

It's probably that you have the _Hide whitespace changes_ option toggled at the top of your review window :)
![image](https://github.com/user-attachments/assets/e974ae77-5aad-4f78-a8ca-21c8448e7f37)

tarekgh commented:
/ba-g the build failure is unrelated
